src/state_machine/state.hpp

```diff
@@ -39,108 +40,58 @@ class Main;  // Forward declaration
 
-  void checkEmergencyStop();
+  virtual void enter(Logger &log) = 0;
+  virtual void exit(Logger &log)  = 0;
 
-  virtual void transitionCheck() = 0;
+  virtual State *checkTransition(Logger &log) = 0;
 
-  Logger &log_;
   data::Data &data_;
 
  protected:
-  Main *state_machine_;
+  static State *instance_;
+  data::EmergencyBrakes embrakes_data_;
+  data::Navigation nav_data_;
+  data::Batteries batteries_data_;
+  data::Telemetry telemetry_data_;
+  data::Sensors sensors_data_;
+  data::Motors motors_data_;
+  void updateModuleData();
 };
 
-class Idling : public State {
- public:
-  Idling(Logger &log, Main *state_machine) : State(log, state_machine) {}
-
-  /*
-   * @brief   Checks for calibration command
-   */
-  void transitionCheck();
-};
-
-class Calibrating : public State {
- public:
-  Calibrating(Logger &log, Main *state_machine) : State(log, state_machine) {}
-
-  /*
-   * @brief   Checks if the calibration has been completed.
-   */
-  void transitionCheck();
-};
-
-class Ready : public State {
- public:
-  Ready(Logger &log, Main *state_machine) : State(log, state_machine) {}
-
-  /*
-   * @brief   Checks for launch command
-   */
-  void transitionCheck();
-};
-
-class Accelerating : public State {
- public:
-  Accelerating(Logger &log, Main *state_machine) : State(log, state_machine) {}
-
-  /*
-   * @brief   Checks for critical failure during run.
-   */
-  void checkEmergencyStop();
-
-  /*
-   * @brief   Checks if max distance reached
-   */
-  void transitionCheck();
-};
-
-class NominalBraking : public State {
- public:
-  NominalBraking(Logger &log, Main *state_machine) : State(log, state_machine) {}
-
-  /*
-   * @brief   Checks for critical failure during run.
-   */
-  void checkEmergencyStop();
-
-  /*
-   * @brief   Checks whether the pod has stopped.
-   */
-  void transitionCheck();
-};
-
-class Finished : public State {
- public:
-  Finished(Logger &log, Main *state_machine) : State(log, state_machine) {}
-
-  /*
-   * @brief   Checks if command to shut down was sent
-   */
-  void transitionCheck();
-};
-
-class FailureBraking : public State {
- public:
-  FailureBraking(Logger &log, Main *state_machine) : State(log, state_machine) {}
-
-  /*
-   * @brief   Cheks whether the pod has stopped.
-   */
-  void transitionCheck();
-};
+/*
+ * @brief   Generates a specific state S following the pattern of State.
+ */
+#define MAKE_STATE(S)                                                                              \
+  class S : public State {                                                                         \
+   public:                                                                                         \
+    S() {}                                                                                         \
+    static S *instance_;                                                                           \
+    static S *getInstance() { return S::instance_; }                                               \
```

Comment:
I think if I only do it in state it will always refer to `State::instance_` and not to `S::instance_`. Because the members are static you can't make them virtual, as far as I understand. So you have to know which types' member you need.

I can check that, though.